### PR TITLE
resolve logic errors introduced in #11077 and return 40x errors instead of 500 errors where possible

### LINKED ISSUE #11736: Triggering build with invalid branch will cause the invoke to fail and return a code 500
Following from #11077, Wang has pointed out the following at https://bugzilla.redhat.com/show_bug.cgi?id=1373330 :

Triggering build with invalid branch will cause the invoke to fail and return a code 500 , internal error,. Should we return code 400, i.e. a bad request?

This is effectively a question of where we want to stand on the API backwards compatibility vs otherwise correctness continuum.

Reviewing the return paths, you could argue that there are points where we should be returning 400 bad request, 401 unauthorized, 404 not found and 500 internal error.  Almost certainly the same would apply to the GitHub trigger.  However today it's all 500s.

NB: this issue is not a regression due to #11077 being merged, it's the same behaviour as before that is being questioned.

@smarterclayton please could you provide a view on whether you want this behaviour to change?

@bparees @wanghaoran1988 fyi

### PR DESCRIPTION
**Pull request by jim-minter**
fixes #11736

### CODE REVIEW

**Comment by bparees** [comment]
@jim-minter i'm failing to see how this addresses the missing branch behavior in the referenced issue?

File: pkg/build/registry/buildconfig/webhook.go [hunk 1]
```diff
@@ -55,11 +55,13 @@ func (w *WebHook) ServeHTTP(writer http.ResponseWriter, req *http.Request, ctx k
 		switch err {
 		case webhook.ErrSecretMismatch, webhook.ErrHookNotEnabled:
 			return errors.NewUnauthorized(fmt.Sprintf("the webhook %q for %q did not accept your secret", hookType, name))
-		case nil:
-			return nil
-		default:
+		case webhook.MethodNotSupported:
+			return errors.NewMethodNotSupported(buildapi.Resource("buildconfighook"), req.Method)
+		}
+		if _, ok := err.(*errors.StatusError); !ok {
```
Comment (bparees) [inline pkg/build/registry/buildconfig/webhook.go]: is the branch issue resolved by getting rid of the "default" from the switch?

Comment (jim-minter) [inline pkg/build/registry/buildconfig/webhook.go]: yes, and then passing err through untouched if it is of type *errors.StatusError.

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
```
--- FAIL: TestWebhookGitHubPing (4.98s)
	webhookgithub_test.go:300: Wrong response code, expecting 200, got 500: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Internal error occurred: hook failed: \u003cnil\u003e","reason":"InternalError","details":{"causes":[{"message":"hook failed: \u003cnil\u003e"}]},"code":500}
```

**Comment by jim-minter** [comment]
Sorry, one missing err != nil check.
[test]

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
looks like network flakes.
[test]

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/10988
flake https://github.com/openshift/origin/issues/10228
flake https://github.com/openshift/origin/issues/10773
flake https://github.com/openshift/origin/issues/9076

(quite an impressive collection)

[test]

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/9886
[merge]

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/10773
[merge]

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/11750
flake https://github.com/openshift/origin/issues/10773

[merge]

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/11662
flake https://github.com/openshift/origin/issues/11024
[merge]

**Comment by bparees** [comment]
[merge]

Ben Parees | OpenShift

**Comment by bparees** [comment]
[test]

**Comment by jim-minter** [comment]
flake #9076 again

**Comment by bparees** [comment]
well that has to be some sort of record.

**Pull request merged**